design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+more featureful tools such as Helm or jsonnet when they hit the limitations
+imposed by the `Composition` type, or even writing their own Kubernetes
+controllers (typically in Go, using [controller-runtime]) to build bespoke
+abstractions. Furthermore, some Crossplane users simply prefer to express their
+intent in a language (or with a tool) they already know and feel comfortable
+with, such as Helm, jsonnet, Python, HCL, Typescript, etc.
+
+While it's true that users _could_ simply use Helm, jsonnet, etc etc to compose
+Crossplane managed resources into higher level abstractions (like the
+aforementioned `AcmeCoDatabase` example), we feel that these abstractions are
+best exposed as _APIs_, not as client-side constructs.
+
+Exposing abstractions as a (Kubernetes) API:
+
+* Avoids the need to distribute (and version) client-side tooling as part of the
+  collaboration process.
+* Allows abstractions to be access controlled using RBAC.
+* Increases tooling support - more tools have native support for calling REST
+  APIs than (e.g.) building Helm charts.
+* _Further_ increases tooling support by leveraging the Kubernetes ecosystem -
+  things like ArgoCD and OPA.
+
+## Goals
+
+The proposal put forward by this document should:
+
+* Let folks use their composition tool and/or programming language of choice.
+* Support 'advanced' composition logic such as loops and conditionals.
+* Balance safety (e.g. sandboxing) with speed and simplicity.
+* Be possible to introduce behind a feature flag that is off by default.
+
+While not an explicit goal, it would also be ideal if the solution put forth by
+this document could serve as a test bed for new features in the contemporary
+'resources array' based form of Composition.
+
+## Proposal
+
+### Overview
+
+This document proposes that a new `functions` array be added to the existing
+`Composition` type. This array of functions would be called either instead of or
+in addition to the existing `resources` array in order to determine how an XR
+should be composed.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: my-cool-function
+    type: Container
+    container:
+      image: xkpg.io/my-cool-function:0.1.0
+```
+
+Under this proposal each function is the entrypoint of an OCI image, though the
+API is designed to support different function implementations (such as webhooks)
+in th future. The updated API would affect only the `Composition` type - no
+changes would be required to the schema of `CompositeResourceDefinitions`, XRs,
+etc.
+
+Notably the functions would not be responsible for interacting with the API
+server to create, update, or delete composed resources. Instead they instruct
+Crossplane which resources should be created, updated, or deleted.
```

Comment:
100% on this, something that I am not liking too much is the fact that Crossplane seems to be owning the lifecycle of these functions. The function definition in the previous example showed the container image for the function and I wouldn't go that far. The composition can check if the functions are available by checking by reference. Check comment above.